### PR TITLE
Changes for HHVM compatibility

### PR DESCRIPTION
**Pull request by nicholas-eden**
https://github.com/briannesbitt/Carbon/issues/427

It looks like later versions will fix some of the bugs, not sure which versions this will pass for besides 3.10.1

### CODE REVIEW

**Comment by nicholas-eden** [comment]
We are down to 1 failing assert for hhvm and it isn't one that can be fixed with a workaround.

There are a few things I'm not too proud of, a few hacks in isHhvm blocks.

File: src/Carbon/Carbon.php [hunk 1]
```diff
@@ -198,6 +200,18 @@ protected static function safeCreateDateTimeZone($object)
         return $tz;
     }
 
+    /**
+     * Determines if code is being executed by HHVM
+     *
+     * @return bool
+     *
+     * @codeCoverageIgnore
```
Comment (lucasmichot) [inline src/Carbon/Carbon.php]: We avoid that `@codeCoverageIgnore` tag.

File: src/Carbon/Carbon.php [hunk 2]
```diff
@@ -772,11 +786,23 @@ public function tz($value)
...
      */
     public function setTimezone($value)
     {
-        return parent::setTimezone(static::safeCreateDateTimeZone($value));
+        try {
+            return parent::setTimezone(static::safeCreateDateTimeZone($value));
+        } catch (Exception $e) {
+            // @codeCoverageIgnoreStart
+            $message = $e->getMessage();
+            if (self::isHhvm() && 0 === strpos($message, 'DateTimeZone::__construct()')) {
```
Comment (lucasmichot) [inline src/Carbon/Carbon.php]: We don't use yoda conditions, `strpos($message, 'DateTimeZone::__construct()') === 0` is to be chosen

**Comment by lucasmichot** [comment]
Just an idea, would it not make sense to make an extended class which only target HHVM...  ~~ warum nicht
@briannesbitt @dereuromark, thoughts ?

**Comment by nicholas-eden** [comment]
Ideally there would be no code changes required when switching engines, so if it were a new class specific to hhvm some form of composition might work.

File: src/Carbon/Carbon.php (hunk 2 above)
Comment (GrahamCampbell) [inline src/Carbon/Carbon.php]: also, should be static:: rather than self::

**Comment by nicholas-eden** [comment]
That last test will be fixed a slightly later version of hhvm: https://github.com/facebook/hhvm/pull/5240

I'm guessing 3.7, not sure how often Travis CI updates hhvm or if @briannesbitt needs to do it.

**Comment by nicholas-eden** [comment]
Nevermind, the answer is here: http://stackoverflow.com/questions/32151774/does-travis-ci-support-hack

Travis is currently locked to 3.6.  We could put in a skip for that assertion based on the hhvm version.  Everything else would pass, allowing hhvm to be removed from the ignore list.

**Comment by lucasmichot** [comment]
Can you rebase @nicholas-eden ?

**Comment by lucasmichot** [comment]
Can you rebase @nicholas-eden ?

**Comment by kylekatarnls** [comment]
Hi,

IMO, we simply should drop HHVM support. HHVM brought so much pain to so many people. It was a good idea and the performance worsted this pain before, but now with the PHP 7 epoch it's tile to let this old good friend rest in peace mostly if it needs many specific code.

I'm sorry for your hard work, it would indeed be relevant in 2015 when  you did it, but I think I would close this PR if anyone else has arguments to oppose.

**Comment by briannesbitt** [comment]
Based on https://hhvm.com/blog/2017/09/18/the-future-of-hhvm.html it looks like they will no longer officially be supporting all of the PHP language features.  I think its best this doesn't get merged and we don't specifically support HHVM moving forward.  This could be done in a subclass if need be in the future, but probably in a different repo.

**Pull request closed without merging**